channel/state/state.go

```diff
@@ -168,21 +169,47 @@ func (s State) RecoverSigner(sig Signature) (types.Address, error) {
 	return RecoverEthereumMessageSigner(stateHash[:], sig)
 }
 
+// equalParticipants returns true if the given arrays contain equal addresses (in the same order).
+func equalParticipants(p []types.Address, q []types.Address) bool {
+	if len(p) != len(q) {
+		return false
+	}
+	for i, a := range p {
+		if !bytes.Equal(a.Bytes(), q[i].Bytes()) {
+			return false
+		}
+	}
+	return true
+}
+
+// Equal returns true if the given State is deeply equal to the receiever.
+func (s State) Equal(r State) bool {
+	return s.ChainId.Cmp(r.ChainId) == 0 &&
+		equalParticipants(s.Participants, r.Participants) &&
+		s.ChannelNonce.Cmp(r.ChannelNonce) == 0 &&
+		bytes.Equal(s.AppDefinition.Bytes(), r.AppDefinition.Bytes()) &&
+		s.ChallengeDuration.Cmp(r.ChallengeDuration) == 0 &&
+		bytes.Equal(s.AppData, r.AppData) &&
+		s.Outcome.Equal(r.Outcome) &&
+		s.TurnNum.Cmp(r.TurnNum) == 0 &&
+		s.IsFinal == r.IsFinal
+}
+
 // Clone returns a clone of the state
 func (s State) Clone() State {
 	clone := State{}
 
-	// shallow clone of fixed part
-	clone.ChainId = s.ChainId
+	// Fixed part
+	clone.ChainId = new(big.Int).Set(s.ChainId)
 	clone.Participants = s.Participants
```

Comment:
```suggestion
	clone.Participants = append(clone.Participants, s.Participants...)
```
See: https://go.dev/play/p/twgDh-4MD3s

Finally it has clicked for me that fixed-length arrays are comparable (`==`) and safely copyable (`safeCopy := origininal`) but slices are not. This jarred with my notion of slices being the universally friendlier container.